libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt

```diff
@@ -0,0 +1,67 @@
+package com.mapbox.navigation.ui.voice.api
+
+import com.mapbox.api.directions.v5.models.VoiceInstructions
+import com.mapbox.navigation.utils.internal.logW
+
+internal class TimeBasedNextVoiceInstructionsProvider(
+    private val observableTimeSeconds: Int,
+) : NextVoiceInstructionsProvider {
+
+    override fun getNextVoiceInstructions(progress: RouteProgressData): List<VoiceInstructions> {
+        val legs = progress.route.legs() ?: return emptyList<VoiceInstructions>().also {
+            logW(
+                "Route does not contain legs, progress=$progress",
+                LOG_CATEGORY
+            )
+        }
+        var legSteps = legs.getOrNull(progress.legIndex)?.steps()
+        var currentStep = legSteps
+            ?.getOrNull(progress.stepIndex)
+            ?: return emptyList<VoiceInstructions>().also {
+                logW(
+                    "Route does not contain valid step, progress=$progress",
+                    LOG_CATEGORY
+                )
+            }
+
+        val voiceInstructions = mutableListOf<VoiceInstructions>()
+        var cumulatedTime = progress.stepDurationRemaining
+        val currentStepInstructions = currentStep.voiceInstructions()?.filter { instruction ->
+            val distanceAlongGeometry = instruction.distanceAlongGeometry()
+            distanceAlongGeometry != null
+                && distanceAlongGeometry <= progress.stepDistanceRemaining
+        }
+        if (currentStepInstructions != null) {
+            voiceInstructions.addAll(currentStepInstructions)
+        }
+
+        var currentStepIndex = progress.stepIndex
+        var currentLegIndex = progress.legIndex
+        while (cumulatedTime < observableTimeSeconds) {
+            if (currentStepIndex + 1 < (legSteps?.size ?: 0)) {
+                currentStep = legSteps!![currentStepIndex + 1]
+                currentStepIndex++
+            } else {
+                currentStepIndex = 0
+                if (currentLegIndex + 1 < legs.size) {
+                    legSteps = legs[currentLegIndex + 1].steps()
+                    currentLegIndex++
+                    if (legSteps.isNullOrEmpty()) {
+                        continue
+                    } else {
+                        currentStep = legSteps.first()
+                    }
+                } else {
+                    break
+                }
+            }
+            currentStep.voiceInstructions()?.let { voiceInstructions.addAll(it) }
+            cumulatedTime += currentStep.duration()
+        }
```

Comment:
Maybe?
```suggestion
        findStepsAhead(observableTimeSeconds - cumulatedTime, progress).forEach {
            currentStep.voiceInstructions()?.let { voiceInstructions.addAll(it) }
        }
```
where
```
fun findStepsAhead(timeAhead: Int, progress: RouteProgressData): List<RouteStep> {
        val result = mutableListOf<RouteStep>()
        var currentStepIndex = progress.stepIndex
        var currentLegIndex = progress.legIndex
        while (cumulatedTime < observableTimeSeconds) {
            if (currentStepIndex + 1 < (legSteps?.size ?: 0)) {
                currentStep = legSteps!![currentStepIndex + 1]
                currentStepIndex++
            } else {
                currentStepIndex = 0
                if (currentLegIndex + 1 < legs.size) {
                    legSteps = legs[currentLegIndex + 1].steps()
                    currentLegIndex++
                    if (legSteps.isNullOrEmpty()) {
                        continue
                    } else {
                        currentStep = legSteps.first()
                    }
                } else {
                    break
                }
            }
            result.add(currentStep)
            cumulatedTime += currentStep.duration()
        }
}
```

Reply:
This code does not compile and I don't really see what you're suggesting.

Reply:
> This code does not compile

sorry for that.

I had difficulties understanding what's going on there. I understood it as following:
1. iteration over steps from different legs until observableTimeSeconds is reached  
2. add all voice instructions from the steps to result

So I decided to suggest extract the iteration logic into a function to simplify readability. Readability is a subjective topic, so all I'm doing is sharing opinion of how to make code more readable for me